MM-11701 Log request ID for outgoing webhooks

JerryFireman opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/10411

mattermod commented:
Thanks @JerryFireman for the pull request!

 Per the [**CONTRIBUTING.md**](https://github.com/mattermost/platform/blob/master/CONTRIBUTING.md#submitting-a-pull-request) file displayed when you created this pull request, we need to add you to the list of [approved contributors](https://docs.google.com/spreadsheets/d/1NTCeG-iL_VS9bFqtmHSfwETo5f-8MQ7oMDE5IUYJi_Y/pubhtml?gid=0&single=true) for the Mattermost project.

**Please help complete the Mattermost [contribution license agreement](http://www.mattermost.org/mattermost-contributor-agreement/)?**

This is a standard procedure for many open source projects. Your form should be processed within 24 hours and reviewers for your pull request will be able to proceed.

Please let us know if you have any questions.

We are very happy to have you join our growing community! If you're not yet a member, please consider joining our [Contributors community channel](https://pre-release.mattermost.com/core/channels/tickets) to meet other contributors and discuss new opportunities with the core team.

JerryFireman commented:
@crspeller I deleted the previous branch and therefore had to close https://github.com/mattermost/mattermost-server/pull/10484.  Sorry for any confusion. I added the request_ID to the incoming webhooks log behind a defer statement to address your comments. I ran the linter and unit tests. Please let me know how it looks and if anything else is missing or can be improved. Thanks very much for all of your help!

JerryFireman commented:
@crspeller @jwilander just checking in for an update? is there anything I need to do on this one?

crspeller commented:
@JerryFireman I don't think this solves the problem. A couple things need to happen:
1. The log message needs to be converted to using structured logging.
2. Move the debug log into a defer so it will run on any error and print the webhook body for debugging purposes. You need to check in the defer what was returned and only log on errors.

JerryFireman commented:
@crspeller thanks for the feedback! I will correct and update the PR.

JerryFireman commented:
@crspeller Can you please let me know where to find the structured log produced by mlog on my local machine? Thanks!

crspeller commented:
@JerryFireman It'a always output in a structured format. But if you want to see the JSON output you can removing the pipe to the formatter in the makefile: https://github.com/mattermost/mattermost-server/blob/master/Makefile#L443

You can change your config to use file logging and the file will be in the JSON format.

JerryFireman commented:
@crspeller I converted the log to structured logging and moved the debug log into a defer that should generate a log when and only when an error occurs. I played with the makefile but was unable to generate a json log. Please let me know how it looks when you have a chance to review.

jasonblais commented:
/check-cla

jasonblais commented:
/cla-check

JerryFireman commented:
@jasonblais I filled in the cla again -- i must have done it wrong the first time. please let me know if there's anything else you need from me.

jasonblais commented:
@cpanato looks like our CLA check is case-sensitive (@JerryFireman filled his CLA with username `jerryfireman`). I believe we've encountered this before.

Would it be easy to update our CLA check not to be case-sensitive (i.e. treat `JerryFireman` same as `jerryfireman`, given GitHub usernames are not case-sensitive?

cpanato commented:
/check-cla

cpanato commented:
/check-cla

cpanato commented:
@jasonblais fixed

JerryFireman commented:
@jasonblais @crspeller just checking in for an update? is there anything I need to do on this one?

jasonblais commented:
Hey @JerryFireman, sorry for the delay. Should be ready for another review from @crspeller

hanzei commented:
@crspeller @jwilander Could you please review when you have the time?

crspeller commented on web/webhook.go:
```diff
@@ -56,9 +55,13 @@ func incomingWebhook(c *Context, w http.ResponseWriter, r *http.Request) {
-	if *c.App.Config().LogSettings.EnableWebhookDebugging {
-		mlog.Debug(fmt.Sprintf("Incoming webhook received. Id=%s Content=%s", id, incomingWebhookPayload.ToJson()))
-	}
+	defer func() {
```
This defer should be at the top of the function. Otherwise it won't do anything. 
You are also checking the wrong error. We want to check if we are returning an error and log it.

JerryFireman replied:
@crspeller Thanks very much for your feedback! I addressed your comments and updated the PR. Please let me know how it looks.

JerryFireman commented:
@crspeller Just checking in for an update, Do you have any feedback for me?

crspeller commented on web/webhook.go:
```diff
@@ -30,6 +29,15 @@ func incomingWebhook(c *Context, w http.ResponseWriter, r *http.Request) {
+
+	defer func() {
+		if *c.App.Config().LogSettings.EnableWebhookDebugging {
+			if err != nil {
```
Probably better to check `c.Err` because that's really the "return" value.

crspeller commented on web/webhook.go:
```diff
@@ -30,6 +29,15 @@ func incomingWebhook(c *Context, w http.ResponseWriter, r *http.Request) {
+	defer func() {
+		if *c.App.Config().LogSettings.EnableWebhookDebugging {
+			if err != nil {
+				mlog.Debug("Incoming webhook received", mlog.String("Id", id), mlog.String("Request ID", c.App.RequestId), mlog.String("Payload", incomingWebhookPayload.ToJson()))
```
Our convention is to use lower snakecase for the keys. So `webhook_id`, `request_id`, `payload` please.

JerryFireman replied:
@crspeller Thanks very much for the feedback. I addressed your comments and pushed a new commit. Please let me know if you see any other areas that need to be addressed.

crspeller commented:
@JerryFireman Your build is failing because you need to run `gofmt`.

JerryFireman commented:
@crspeller Oops, sorry I missed that. Ran gofmt and tests all pass now.
